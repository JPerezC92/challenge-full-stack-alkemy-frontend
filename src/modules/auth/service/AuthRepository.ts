import { UserRegister } from "src/modules/auth/dto/UserRegister";
import { AccessCredentials } from "src/modules/auth/models/AccessCredentials";
import { UserLogin } from "../dto/UserLogin";

export interface AuthRepository {
  register: (userCreate: UserRegister) => Promise<AccessCredentials | void>;
  login: (userLogin: UserLogin) => Promise<AccessCredentials | void>;
}
