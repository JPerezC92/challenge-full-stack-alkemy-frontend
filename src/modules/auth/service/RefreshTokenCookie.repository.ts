import Cookies from "js-cookie";

export function RefreshTokenCookieRepository() {
  const cookieName = "refreshToken";
  return {
    save: (refreshToken: string): void => {
      Cookies.set(cookieName, refreshToken, {
        expires: 1,
        path: "/",
        sameSite: "strict",
        secure: true,
      });
    },
    get: (): string => Cookies.get(cookieName) || "",
  };
}
