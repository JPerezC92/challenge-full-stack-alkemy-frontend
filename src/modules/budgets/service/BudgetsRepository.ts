export interface BudgetsRepository {
  getBalance(): Promise<number | undefined>;
}
