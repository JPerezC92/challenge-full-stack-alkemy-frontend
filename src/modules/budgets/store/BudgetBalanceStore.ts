export interface BudgetBalanceStore {
  updateBalance(balance: number): void;
}
