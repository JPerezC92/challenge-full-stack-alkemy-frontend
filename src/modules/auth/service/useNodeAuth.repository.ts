import React from "react";
import { toast } from "react-toastify";
import { AuthLoginPostEndpoint } from "src/modules/auth/dto/AuthLoginPostEndpoint";
import { AuthRefreshTokenGetEndpoint } from "src/modules/auth/dto/AuthRefreshTokenGetEndpoint";
import { AuthRegisterPostEndpoint } from "src/modules/auth/dto/AuthRegisterPostEndpoint";
import { UserEndpointToDomain } from "src/modules/auth/mappers/UserEndpointToDomain";
import { AccessCredentials } from "src/modules/auth/models/AccessCredentials";
import { JsendStatus } from "src/modules/shared/service/JsendResponse";
import { MyRepository } from "src/modules/shared/service/MyRepository";
import { RequestMethod } from "src/modules/shared/service/RequestMethod";
import { BASE_API_URL } from "src/modules/shared/utils/constants";
import { formatBearerToken } from "src/modules/shared/utils/formatBearerToken";
import { AuthLogoutTokenGetEndpoint } from "../dto/AuthLogoutTokenGetEndpoint";
import { AuthRepository } from "./AuthRepository";
import { RefreshTokenCookieRepository } from "./RefreshTokenCookie.repository";

export function useNodeAuthRepository(): MyRepository<AuthRepository> {
  return React.useCallback(({ abortController }) => {
    const authApiUrl = `${BASE_API_URL}/auth`;
    const refreshTokenCookieRepository = RefreshTokenCookieRepository();

    return {
      register: async ({
        firstName,
        lastName,
        email,
        password,
      }): Promise<AccessCredentials | void> => {
        const response = await fetch(`${authApiUrl}/register`, {
          body: JSON.stringify({ firstName, lastName, email, password }),
          headers: { "Content-Type": "application/json" },
          method: RequestMethod.POST,
          signal: abortController.signal,
        });

        const result = (await response.json()) as AuthRegisterPostEndpoint;

        if (result.status !== JsendStatus.success) {
          toast.warning(result.message, { autoClose: 5000 });
          return;
        }

        const { accessToken, refreshToken, user } = result.data;

        refreshTokenCookieRepository.save(refreshToken);

        return { user: UserEndpointToDomain(user), accessToken: accessToken };
      },

      login: async (userLogin): Promise<AccessCredentials | void> => {
        const response = await fetch(`${authApiUrl}/login`, {
          body: JSON.stringify({ ...userLogin }),
          headers: { "Content-Type": "application/json" },
          method: RequestMethod.POST,
          signal: abortController.signal,
        });

        const result = (await response.json()) as AuthLoginPostEndpoint;

        if (result.status !== JsendStatus.success) {
          toast.warning(result.message, { autoClose: 5000 });
          return;
        }

        const { accessToken, refreshToken, user } = result.data;

        refreshTokenCookieRepository.save(refreshToken);

        return { accessToken, user: UserEndpointToDomain(user) };
      },

      logout: async ({ accessToken }) => {
        const response = await fetch(`${authApiUrl}/logout`, {
          headers: { Authorization: formatBearerToken(accessToken) },
          signal: abortController.signal,
          method: RequestMethod.GET,
        });

        const result = (await response.json()) as AuthLogoutTokenGetEndpoint;

        if (result.status !== JsendStatus.success) {
          toast.warning(result.message, { autoClose: 5000 });
          return;
        }

        refreshTokenCookieRepository.remove();

        return true;
      },

      refreshToken: async (): Promise<AccessCredentials | void> => {
        const refreshTokenCookie = refreshTokenCookieRepository.get();

        if (!refreshTokenCookie) return;

        const response = await fetch(`${authApiUrl}/refresh-token`, {
          headers: {
            "Content-Type": "application/json",
            "x-refresh-token": formatBearerToken(refreshTokenCookie),
          },
          method: RequestMethod.GET,
          signal: abortController.signal,
        });

        const result = (await response.json()) as AuthRefreshTokenGetEndpoint;

        if (result.status !== JsendStatus.success) return;

        const { accessToken, refreshToken, user } = result.data;

        refreshTokenCookieRepository.save(refreshToken);

        return { accessToken, user: UserEndpointToDomain(user) };
      },
    };
  }, []);
}
