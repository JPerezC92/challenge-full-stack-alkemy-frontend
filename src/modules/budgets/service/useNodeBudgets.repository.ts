import React from "react";
import { useCredentialsState } from "src/modules/auth/containers/PrivateRoute/CredentialsProvider.context";
import { BalanceEndpoint } from "src/modules/budgets/dto/BalanceEndpoint";
import { MyRepository } from "src/modules/shared/service/MyRepository";
import { BASE_API_URL } from "src/modules/shared/utils/constants";
import { formatBearerToken } from "src/modules/shared/utils/formatBearerToken";
import { BudgetsRepository } from "./BudgetsRepository";

export function useNodeBudgetsRepository(): MyRepository<BudgetsRepository> {
  const { accessToken } = useCredentialsState();

  return React.useCallback(
    ({ abortController }) => {
      const budgetsApiUrl = `${BASE_API_URL}/budgets`;

      return {
        getBalance: async (): Promise<number | undefined> => {
          const response = await fetch(`${budgetsApiUrl}/balance`, {
            signal: abortController.signal,
            method: "GET",
            headers: { Authorization: formatBearerToken(accessToken) },
          });

          const balanceEndpoint = (await response.json()) as BalanceEndpoint;

          return balanceEndpoint.data.balance;
        },
      };
    },
    [accessToken]
  );
}
