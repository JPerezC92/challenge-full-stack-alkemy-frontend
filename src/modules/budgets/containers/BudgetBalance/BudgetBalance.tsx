import React from "react";
import { BudgetsRepository } from "src/modules/budgets/service/BudgetsRepository";
import { BudgetBalanceStore } from "src/modules/budgets/store/BudgetBalanceStore";
import { useRequest } from "src/modules/shared/hooks/useRequest";
import { MyRepository } from "src/modules/shared/service/MyRepository";
import { MyStore } from "src/modules/shared/store/MyStore";

type BudgetBalanceProps = {
  balance: number;
  budgetsRepository: MyRepository<BudgetsRepository>;
  budgetBalanceStore: MyStore<BudgetBalanceStore>;
};

export const BudgetBalance: React.FC<BudgetBalanceProps> = ({
  balance,
  budgetBalanceStore,
  budgetsRepository,
}) => {
  useRequest(
    async ({ abortController }) => {
      const _budgetsRepository = budgetsRepository({ abortController });
      const _budgetBalanceStore = budgetBalanceStore();
      const balance = await _budgetsRepository.getBalance();

      if (!balance) return;

      _budgetBalanceStore.updateBalance(balance);
    },

    [budgetBalanceStore, budgetsRepository]
  );

  return (
    <>
      <h1>Current balance: {balance}</h1>
    </>
  );
};
