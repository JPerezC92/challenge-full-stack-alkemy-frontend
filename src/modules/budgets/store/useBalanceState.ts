import React from "react";

import { MyStore } from "src/modules/shared/store/MyStore";
import { BudgetBalanceStore } from "./BudgetBalanceStore";

export interface BudgetBalanceState
  extends ReturnType<typeof useBudgetBalanceState> {}

export const useBudgetBalanceState = () => {
  const [balance, setBalance] = React.useState(0);

  const budgetBalanceStore = React.useRef<BudgetBalanceStore>({
    updateBalance: (balance: number) => setBalance(balance),
  });

  return {
    balance: balance.toFixed(2),
    budgetBalanceStore: budgetBalanceStore.current,
  };
};
