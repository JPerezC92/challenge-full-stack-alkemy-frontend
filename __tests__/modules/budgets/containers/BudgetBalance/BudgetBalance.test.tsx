import { render, renderHook, waitFor } from "@testing-library/react";
import renderer from "react-test-renderer";
import { BudgetBalance } from "src/modules/budgets/containers/BudgetBalance";
import { BudgetsRepository } from "src/modules/budgets/service/BudgetsRepository";
import * as useNodeBudgetsRepository from "src/modules/budgets/service/useNodeBudgets.repository";
import { BudgetBalanceStore } from "src/modules/budgets/store/BudgetBalanceStore";
import * as useBudgetBalanceState from "src/modules/budgets/store/useBalanceState";

const balance = 1000;

const budgetBalanceStore: BudgetBalanceStore = {
  updateBalance: jest.fn(),
};

const budgetsRepository: BudgetsRepository = {
  getBalance: jest.fn().mockResolvedValue(balance),
};

jest
  .spyOn(useNodeBudgetsRepository, "useNodeBudgetsRepository")
  .mockReturnValue(jest.fn().mockReturnValue(budgetsRepository));

jest.spyOn(useBudgetBalanceState, "useBudgetBalanceState").mockReturnValue({
  balance,
  budgetBalanceStore: () => budgetBalanceStore,
});

describe("BudgetBalance component", () => {
  test("should call getBalance and updateBalance", async () => {
    const budgetBalanceState = renderHook(() =>
      useBudgetBalanceState.useBudgetBalanceState()
    );

    const component = render(
      <BudgetBalance
        balance={budgetBalanceState.result.current.balance}
        budgetBalanceStore={
          budgetBalanceState.result.current.budgetBalanceStore
        }
        budgetsRepository={useNodeBudgetsRepository.useNodeBudgetsRepository()}
      />
    );

    await waitFor(async () => {
      expect(await component.findByText(/100/i)).toBeInTheDocument();
    });

    expect(budgetsRepository.getBalance).toHaveBeenCalledTimes(1);
    expect(budgetBalanceStore.updateBalance).toHaveBeenCalledTimes(1);
    expect(budgetBalanceStore.updateBalance).toHaveBeenCalledWith(balance);
  });

  test("should match the snapshot", async () => {
    const budgetBalanceState = renderHook(() =>
      useBudgetBalanceState.useBudgetBalanceState()
    );

    const component = renderer
      .create(
        <BudgetBalance
          balance={budgetBalanceState.result.current.balance}
          budgetBalanceStore={
            budgetBalanceState.result.current.budgetBalanceStore
          }
          budgetsRepository={useNodeBudgetsRepository.useNodeBudgetsRepository()}
        />
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
