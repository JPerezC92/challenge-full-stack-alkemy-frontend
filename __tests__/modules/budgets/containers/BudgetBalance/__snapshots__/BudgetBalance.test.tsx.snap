// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetBalance component should match the snapshot 1`] = `
<h1>
  Current balance: 
  1000
</h1>
`;
